doc/subframes/src/subframes_tutorial.cpp

```diff
@@ -210,23 +251,39 @@ int main(int argc, char** argv)
   planning_scene_interface.applyAttachedCollisionObject(att_coll_object);
   // END_SUB_TUTORIAL
 
+  // Fetch the current planning scene state once
+  auto planning_scene_monitor = std::make_shared<planning_scene_monitor::PlanningSceneMonitor>("robot_description");
+  planning_scene_monitor->requestPlanningSceneState();
+  planning_scene_monitor::LockedPlanningSceneRO planning_scene(planning_scene_monitor);
+
+  // Visualize frames as rviz markers
+  ros::Publisher marker_publisher = nh.advertise<visualization_msgs::MarkerArray>("visualization_marker_array", 10);
+  auto showFrames = [&](geometry_msgs::PoseStamped target, const std::string& eef) {
+    visualization_msgs::MarkerArray markers;
+    // convert target pose into planning frame
+    Eigen::Isometry3d tf;
+    tf2::fromMsg(target.pose, tf);
+    target.pose = tf2::toMsg(planning_scene->getFrameTransform(target.header.frame_id) * tf);
+    target.header.frame_id = planning_scene->getPlanningFrame();
+    createFrameMarkers(markers, target, "target");
+
+    // convert eef in pose relative to panda_hand
+    target.header.frame_id = "panda_hand";
+    target.pose = tf2::toMsg(planning_scene->getFrameTransform(target.header.frame_id).inverse() *
+                             planning_scene->getFrameTransform(eef));
+    createFrameMarkers(markers, target, "eef", true);
+
+    marker_publisher.publish(markers);
+  };
+
   // Define a pose in the robot base.
-  tf2::Quaternion orientation, orientation_1, target_orientation;
-  geometry_msgs::PoseStamped fixed_pose, temp_pose_stamped;
+  tf2::Quaternion target_orientation;
+  geometry_msgs::PoseStamped fixed_pose, target_pose;
   fixed_pose.header.frame_id = "panda_link0";
   fixed_pose.pose.position.y = -.4;
   fixed_pose.pose.position.z = .3;
-  orientation.setRPY(0, (-20.0 / 180.0 * M_PI), 0);
-  fixed_pose.pose.orientation = tf2::toMsg(orientation);
-
-  // BEGIN_SUB_TUTORIAL quaternions1
```

Comment:
I like how clean the file is now, and there must be a better place for this quaternion stuff, but I'm sure that a refresher on how to handle rotations with TF will really help some newbies out here. Do you have any good materials to link? Worst case, a comment like this may help: "Note that you can split up your rotations into intermediate steps and multiply `tf2 quaternions<http://docs.ros.org/api/tf2/html/classtf2_1_1Quaternion.html>`_ to create your desired orientation more easily."

Reply:
I am using http://wiki.ros.org/agni_tf_tools to interactively setup my rotations.

Reply:
In my opinion that's definitely worth linking. It's not easy for newcomers to find these kinds of tools.